### ISSUE TITLE
Stochasticity for DistributedDataParallel on CPU but not on GPU

### ISSUE THREAD

**Issue by gmoneymiller** [pytorch/pytorch#36871]
## 🐛 Bug

When I run my DDP model on the GPU using a specific seed, I get identical results time after time.  But when I run the same DDP model on the CPU using the same seed, I get different results each time.  

I am setting the seed within the child process, near its beginning, using the following code:

    torch.manual_seed(args.seed)
    np.random.seed(args.seed)

I also tried setting the seed inside the model, before it goes into its training loop.  No luck.

## To Reproduce

1.Run the model via DDP on the GPU using a specific seed, three times, saving the output into different files each time.  Confirm the output is identical across all three files.
2.Run the same model via DDP on the CPU using a specific seed, three times, saving the output into different files each time.  Confirm the output is different across all three files.

## Expected behavior

The output should be the same each time for a given seed, on a given device.  I do not expect CPU and GPU output to match each other, but I do expect all GPU output to match itself, and the same for all CPU output.

## Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 418.74
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.2

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.4.0
[conda] numpy                     1.17.2           py36h95a1406_0    conda-forge

**Comment by pritamdamania87** [comment]
@gmoneymiller Could you provide the complete code to reproduce the problem? I tried a very simple test and couldn't reproduce the problem:

```
import os
import torch
import torch.distributed as dist
from torch.multiprocessing import Process
from torch.nn.parallel import DistributedDataParallel as DDP


def run(rank, size):
    """ Distributed function to be implemented later. """
    torch.manual_seed(42)
    m = torch.nn.Linear(10, 1)
    ddp = DDP(m)
    inp = torch.rand(10)
    out = ddp(inp)
    out.sum().backward()
    print ('OUTPUT: {}'.format(out))
    print ('GRADIENT: {}'.format(m.weight))


def init_process(rank, size, fn, backend='gloo'):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29500'
    dist.init_process_group(backend, rank=rank, world_size=size)
    fn(rank, size)


if __name__ == "__main__":
    size = 2
    processes = []
    for rank in range(size):
        p = Process(target=init_process, args=(rank, size, run))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

Three runs of the script:

```
$ python /tmp/seed.py
OUTPUT: tensor([0.9201], grad_fn=<AddBackward0>)
OUTPUT: tensor([0.9201], grad_fn=<AddBackward0>)
GRADIENT: Parameter containing:
tensor([[ 0.2418,  0.2625, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
          0.2788, -0.2320]], requires_grad=True)
GRADIENT: Parameter containing:
tensor([[ 0.2418,  0.2625, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
          0.2788, -0.2320]], requires_grad=True)
$ python /tmp/seed.py
OUTPUT: tensor([0.9201], grad_fn=<AddBackward0>)
OUTPUT: tensor([0.9201], grad_fn=<AddBackward0>)
GRADIENT: Parameter containing:
tensor([[ 0.2418,  0.2625, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
          0.2788, -0.2320]], requires_grad=True)
GRADIENT: Parameter containing:
tensor([[ 0.2418,  0.2625, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
          0.2788, -0.2320]], requires_grad=True)
$ python /tmp/seed.py
OUTPUT: tensor([0.9201], grad_fn=<AddBackward0>)
OUTPUT: tensor([0.9201], grad_fn=<AddBackward0>)
GRADIENT: Parameter containing:
tensor([[ 0.2418,  0.2625, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
          0.2788, -0.2320]], requires_grad=True)
GRADIENT: Parameter containing:
tensor([[ 0.2418,  0.2625, -0.0741,  0.2905, -0.0693,  0.0638, -0.1540,  0.1857,
          0.2788, -0.2320]], requires_grad=True)
```

**Comment by gmoneymiller** [comment]
Hi!  I will ask permission to share my code with you.

--Scott

**Comment by gmoneymiller** [comment]
I packaged for you a reproducible test case.  Unzip the attached into <TOP_DIR>, and then:

**Comment by gmoneymiller** [comment]
[nail.zip](https://github.com/pytorch/pytorch/files/4526332/nail.zip)
cd <TOP_DIR>/nail/hnn/sho
export PYTHONPATH=$PYTHONPATH:<TOP_DIR>

and then run ./bugtestC.sh (CPU) and ./bugtestG.sh (GPU) several times.  The GPU version takes only a minute or two to run, and the CPU version 5-10 min. 

Each time it will produce a .tsv output file, which are the results of training two models and comparing their performance.  Compare the .tsv files between runs.  You will see that the GPU output is identical each time, but the CPU output differs.

**Comment by gmoneymiller** [comment]
cd <TOP_DIR>/nail/hnn/sho
export PYTHONPATH=$PYTHONPATH:<TOP_DIR>

and then run ./bugtestC.sh (CPU) and ./bugtestG.sh (GPU) several times. The GPU version takes only a minute or two to run, and the CPU version 5-10 min.

Each time it will produce a .tsv output file, which are the results of training two models and comparing their performance. Compare the .tsv files between runs. You will see that the GPU output is identical each time, but the CPU output differs.

**Comment by gmoneymiller** [comment]
Source code package is attached.

--Scott

**Comment by pritamdamania87** [comment]
@gmoneymiller Where can I find this zip file? I don't see a file attached to this issue.

**Comment by gmoneymiller** [comment]
Attached.

--Scott

**Comment by pritamdamania87** [comment]
@gmoneymiller Thanks for sharing the code, I'm able to reproduce this issue locally. I was wondering if you tried running the model multiple times without DDP and noticed a similar issue with determinism? Or does this only happen when you use DDP?

**Comment by pritamdamania87** [comment]
I hacked up the code to use world_size = 1 and it looks like the results are deterministic on CPU. Based on this it seems like the non-determinism is coming from something in DDP.

@mrshenli @zhaojuanmao I was wondering if you had some initial thoughts here in terms of what might be causing this non-determinism? Since this happens only on CPU, one possibility might be that there is some non-determinism in how gloo computes the allreduce result. Another option is that there is some non-determinism in DDP in general, but it only manifests in CPUs for some reason.

**Comment by gmoneymiller** [comment]
I now have what I call a "basic" branch of my code from which I removed all
trace of DDP.  The basic code just does normal training on the CPU, by
default, or on a specific GPU, but it is otherwise the same code as before.

The basic version produces deterministic results on both CPU and GPU.  That
fact does not rule out a bug in my code, but it does strongly suggest the
issue lies within DDP.

Additionally, and this may or may not be related, the results I get from
DDP on the GPU look good and are fit for publication, but on the CPU
the results are terrible.  Same training data, same code, whack results.

Also I notice that the GPU version is slower than the CPU version, which is
counterintuitive.   It happens for both distributed and normal training, so
I'm not sure if it's related to DDP.

I don't want to muddy this bug report, so if you prefer, I can open
separate reports for the preceding two issues.

--Scott

**Comment by zhaojuanmao** [comment]
@pritamdamania where did you find the attached files in the issue? I did not find any link

@gmoneymiller based on what you described, I would like to check whether the CPU codes have something suspicious. Your CPU training results (less accurate, not deterministic, much slower) are completely different from what we observed in our benchmark tool.

**Comment by gmoneymiller** [comment]
Hi Yanli.  Please find attached my code.

--Scott

**Comment by mrshenli** [comment]
@zhaojuanmao the attachment can be downloaded from the email in your inbox with title "[pytorch/pytorch] Stochasticity for DistributedDataParallel on CPU but not on GPU (#36871)"

**Comment by zhaojuanmao** [comment]
Somehow I did not see attachment in the email ☹

**Comment by zhaojuanmao** [comment]
@gmoneymiller I roughly read the attached codes, I saw you wrapped the model using DDP, but did not see the ddp_model forward is actually called. Looks like the codes are still calling the original model forward?

**Comment by rgommers** [comment]
Hi @gmoneymiller, it may be useful to put this code in a gist or (if it's too large for a gist) in a standalone repo. This issue is not actionable without it, and won't be understandable in the future even if it's fixed based on code-in-inbox.

**Comment by gmoneymiller** [comment]
Hi. Sorry for the late reply. I'm a little under the weather. I will
respond as soon as I can.

Thanks,
Scott

**Comment by Baranowski** [comment]
@gmoneymiller, any chance you can share the code that reproduces this issue?

**Comment by gmoneymiller** [comment]
Attached.

Thanks,
Scott

**Comment by gmoneymiller** [comment]
Hi,

That would explain the strange behavior.  What is the correct idiom for
calling the DDP forward?

--Scott

**Comment by gmoneymiller** [comment]
Hi,

What's "gist"?

Thanks,
Scott

**Comment by Baranowski** [comment]
@gmoneymiller, whenever you reply through email, GitHub strips your attachments. Can you please upload/publish your code somewhere and share a link?

**Comment by rgommers** [comment]
>  What's "gist"?

A gist is a public script that's versioned by GitHub: https://gist.github.com/. It's good for these kinds of reproducers and build logs.

**Comment by zhaojuanmao** [comment]
you can call ddp like this:

ddp_model = DistributedDataParallel(model)
output = ddp_model(input)

**Comment by gmoneymiller** [comment]
Thanks for the tip. Will do shortly.

**Comment by gmoneymiller** [comment]
Thanks.  It was my understanding, based on the tutorial examples, that the
model object is modified in-place by the DistributedDataParallel class, so
that one may reference either model or ddp_model.  In other words:

output = model(input)

is the same as

output = ddp_model(input)

Is that not correct?

--Scott

**Comment by gmoneymiller** [comment]
Attached is the zip file containing the code to reproduce the issue.
[nail.zip](https://github.com/pytorch/pytorch/files/4718949/nail.zip)

**Comment by Baranowski** [comment]
@gmoneymiller, have you tried `output = ddp_model(input)`, as suggested by zhaojuanmao?

For some reason I'm getting errors when running GPU version of your script and it would be great to know whether the issue still persists before I spend more time trying to reproduce it.

**Comment by gmoneymiller** [comment]
I am certain the issue persists.  Are you running the test script bugtestG.sh?  What's the error?

--Scott

**Comment by Baranowski** [comment]
I got it working. There was an assumption in a couple places in the code that there are at least 4 GPUs available (I have only 2 on my dev machine).

Anyway, I'm getting different results between runs both from GPU and from CPU. Both on the current master and on torch 1.4.0.

**Comment by gmoneymiller** [comment]
I get consistent results on GPU.  You can adjust the number of GPUs with the command-line parameter --num_gpus.

**Comment by Baranowski** [comment]
Since I cannot reproduce deterministic behaviour with GPU, I'm not in a good position to figure out where the CPU and GPU diverge and ultimately fix the nondeterminism.
